### PR TITLE
feat(fluid-build): support tsc -b

### PR DESCRIPTION
**Pull request by curtisman**
- Check project references with `tsc -b`
- Avoid creating an empty `tsBuildInfo` file when there are no input, by skip checking when checking if the task is up to date.  This allows the project references check work automatically.
- `Webpack` will always get executed if it's dependencies are out of date (`recheckLeafIsUpToDate` is false). No need to include dependencies information in the done file.
- `CopyfilesTask` knows all the input/output file and can enable `recheckLeafIsUpToDate` so that we don't need to trigger it if the dependent task's execution didn't change any of `copyfiles` input.

### CODE REVIEW

File: build-tools/packages/build-tools/src/fluidBuild/buildGraph.ts [hunk 1]
```diff
@@ -518,7 +518,8 @@ export class BuildGraph {
 	}
 
 	public async build(timer?: Timer): Promise<BuildResult> {
-		// TODO: This function can only be called once
+		// This function is only be called once here at the beginning of the build.
+		// It checks the up-to-date state at this moment and will not be changed for the duration of the build.
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/fluidBuild/buildGraph.ts:522]:
```suggestion
		// This function must only be called once here at the beginning of the build.
		// It checks the up-to-date state at this moment and will not be changed for the duration of the build.
```

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts [hunk 2]
```diff
@@ -45,7 +45,12 @@ export abstract class LeafTask extends Task {
 	private _parentLeafTasks: Set<LeafTask> | undefined | null;
 	private parentWeight = -1;
 
-	constructor(node: BuildPackage, command: string, taskName: string | undefined) {
+	constructor(
+		node: BuildPackage,
+		command: string,
+		taskName: string | undefined,
+		private readonly isTemp: boolean = false, // indicate if the task is for temporary and not for execution.
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:52]: What is a temp task? How does it differ from a "group task" or "virtual task" (can't remember what you're calling them exactly) - the ones that don't execute anything themselves, they're just nodes in the graph with children tasks. How are these temp tasks different?

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/webpackTask.ts [hunk 3]
```diff
@@ -7,48 +7,38 @@ import * as fs from "fs";
...
 		try {
 			const config = await loadModule(this.configFileFullPath, this.package.packageJson.type);
 			const content: DoneFileContent = {
 				version: await this.getVersion(),
 				config: typeof config === "function" ? config(this.getEnvArguments()) : config,
 				sources: {},
-				dependencies: {},
 			};
 
+			// TODO: this is specific to repo set up.
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/webpackTask.ts:34]:
```suggestion
			// TODO: this is specific to the microsoft/FluidFramework repo set up.
```

**Comment by tylerbutler** [comment]
> CopyfilesTask knows all the input/output file and can enable recheckLeafIsUpToDate so that we don't need to trigger it if the dependent task's execution didn't change any of copyfiles input.

Can you point me to this change?

**Comment by curtisman** [comment]
> > CopyfilesTask knows all the input/output file and can enable recheckLeafIsUpToDate so that we don't need to trigger it if the dependent task's execution didn't change any of copyfiles input.
> 
> Can you point me to this change?

it is in `miscTasks.ts`

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts (hunk 2 above)
Comment (curtisman) [reply build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:52]: I temporary create `TscTask` for project references just so I can reuse the code to check if it is up-to-date.  It is never entered into the build graph and never executed.   It probably can be refactored better, but again don't want to churn the code.

Comment (tylerbutler) [reply build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:52]: Got it. I think the thing I don't like is that it's not localized to the tsc task (and probably can't be without more refactoring - understood).

**Review by tylerbutler — APPROVED** [review]

**Comment by tylerbutler** [comment]
> > > CopyfilesTask knows all the input/output file and can enable recheckLeafIsUpToDate so that we don't need to trigger it if the dependent task's execution didn't change any of copyfiles input.
> > 
> > 
> > Can you point me to this change?
> 
> it is in `miscTasks.ts`

Ahh, thank you. :)

**Pull request merged**